src/python/pants/engine/internals/graph.py

```diff
@@ -260,21 +282,44 @@ async def resolve_target_parametrizations(
             else:
                 overrides = overrides_field.flatten()
 
-        generated = await Get(
-            GeneratedTargets,
-            GenerateTargetsRequest,
-            generate_request(
-                target,
-                template_address=address,
-                template=template_fields,
-                overrides=overrides,
-            ),
+        all_generated = await MultiGet(
+            Get(
+                GeneratedTargets,
+                GenerateTargetsRequest,
+                generate_request(
+                    target,
+                    template_address=address,
+                    template=template,
+                    overrides={
+                        name: dict(Parametrize.expand(address, override))
+                        for name, override in overrides.items()
+                    },
+                ),
+            )
+            for address, template in Parametrize.expand(address, template_fields)
+        )
+        generated = FrozenDict(
+            GeneratedTargets(
+                target, (t for generated_batch in all_generated for t in generated_batch.values())
+            )
         )
     else:
-        target = target_type(target_adaptor.kwargs, address, union_membership)
-        generated = GeneratedTargets(target, ())
+        first, *rest = Parametrize.expand(address, target_adaptor.kwargs)
+        if rest:
+            target = None
+            generated = FrozenDict(
+                (
+                    parameterized_address,
+                    target_type(parameterized_fields, parameterized_address, union_membership),
+                )
+                for parameterized_address, parameterized_fields in (first, *rest)
+            )
+        else:
+            target = target_type(target_adaptor.kwargs, address, union_membership)
+            generated = FrozenDict()
 
-    for field_type in target.field_types:
+    # TODO: Move to Target constructor.
```

Comment:
Nack, we need to use the engine for singleton so that we don't warn hundreds of times.

Reply:
Yea... but we could use the `@memoized` infrastructure for that, and it would cause it to be triggered in all cases, including for generated targets. Right now it's only triggered for generator targets, and only for non-moved fields.